pkg/collector/corechecks/oracle-dbm/statements.go

```diff
@@ -398,51 +452,96 @@ func (c *Check) StatementMetrics() (int, error) {
 				continue
 			}
 
+			startSQLText := time.Now()
+			queryRow := QueryRow{}
 			var queryHashCol string
-			//if statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature == 0 {
-			if statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature == "" {
-				queryHashCol = "sql_id"
-			} else {
-				queryHashCol = "force_matching_signature"
-			}
-			p := map[string]interface{}{
-				"force_matching_signature": statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature,
-				"sql_id":                   statementMetricRow.StatementMetricsKeyDB.SQLID,
-			}
-			SQLTextQuery := fmt.Sprintf("SELECT sql_fulltext FROM v$sqlstats WHERE %s=:%s AND rownum = 1", queryHashCol, queryHashCol)
+			var SQLStatement string
 
-			rows, err := c.db.NamedQuery(SQLTextQuery, p)
-			if err != nil {
-				log.Errorf("query metrics statements error named exec %s ", err)
-				SQLTextErrors++
-				continue
+			found := false
+			if statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature != "" {
+				obfuscatedStatement, ok := c.statementsCache.forceMatchingSignatures[statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature]
+				if ok {
+					queryRow.Commands = obfuscatedStatement.commands
+					queryRow.QuerySignature = obfuscatedStatement.querySignature
+					queryRow.Tables = obfuscatedStatement.tables
+					SQLStatement = obfuscatedStatement.statement
+					found = true
+				} else if c.config.InstanceConfig.QueryMetrics.IncludeDatadogQueries {
+					obfuscatedStatement, ok := c.DDPrevStatementsCache.forceMatchingSignatures[statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature]
```

Comment:
The "regular" cache, i.e. `c.statementsCache`, is filled by query samples. The agent is collecting query metrics for the queries captured in two consecutive samples. Since the agent statements are excluded from sampling, they would never land in this cache. For this reason, I introduced a separate cache for the agent statements.

Reply:
> Since the agent statements are excluded from sampling 

are all of the agent statements excluded from statements or just the load sampling query? It feels like we _should_ probably collect all of the agent statement that are _not_ the current session aka the sampling query as part of the activity sampling, no?